scripts/grid.js

```diff
@@ -0,0 +1,386 @@
+
+    /**
+     * @type {CameraComponent}
+     * @private
+     */
+    _camera;
+
+    /**
+     * @type {QuadRender}
+     * @private
+     */
+    _quadRender;
+
+    /**
+     * @type {BlendState}
+     */
+    _blendState = new BlendState(
+        true,
+        BLENDEQUATION_ADD, BLENDMODE_SRC_ALPHA, BLENDMODE_ONE_MINUS_SRC_ALPHA,
+        BLENDEQUATION_ADD, BLENDMODE_ONE, BLENDMODE_ONE_MINUS_SRC_ALPHA
+    );
+
+    /**
+     * @type {() => void}
+     */
+    _prerender;
+
+    /**
+     * @type {Vec2}
+     */
+    _halfExtents = new Vec2(Infinity, Infinity);
+
+    /**
+     * @type {Color}
+     * @private
+     */
+    _colorX = new Color(1, 0.3, 0.3);
+
+    /**
+     * @type {Color}
+     * @private
+     */
+    _colorZ = new Color(0.3, 0.3, 1);
+
+    /**
+     * @type {string}
+     */
+    layerName = 'World';
+
+    /**
+     * @param {object} args - The script arguments.
+     */
+    constructor(args) {
+        super(args);
+
+        this._device = this.app.graphicsDevice;
+
+        // create shader
+        const shader = createShaderFromCode(this._device, vertexShader, fragmentShader, 'grid', {
+            vertex_position: SEMANTIC_POSITION
+        });
+        this._quadRender = new QuadRender(shader);
+
+        // set initial size
+        this._set('half_extents', this._halfExtents);
+
+        // set initial colors
+        this._set('color_x', this._colorX);
+        this._set('color_z', this._colorZ);
+    }
+
+    /**
+     * Set the value of a uniform in the shader.
+     *
+     * @param {string} name - The name of the uniform.
+     * @param {Color|Vec3|number} value - The value to set.
+     * @private
+     */
+    _set(name, value) {
+        if (value instanceof Color) {
+            this._device.scope.resolve(name).setValue([value.r, value.g, value.b]);
+        }
+
+        if (value instanceof Vec3) {
+            this._device.scope.resolve(name).setValue([value.x, value.y, value.z]);
+        }
+
+        if (value instanceof Vec2) {
+            this._device.scope.resolve(name).setValue([value.x, value.y]);
+        }
+
+    }
+
+    /**
+     * @attribute
+     * @type {Vec2}
+     */
+    set halfExtents(value) {
+        this._halfExtents.copy(value);
+        this._set('half_extents', this._halfExtents);
+    }
+
+    get halfExtents() {
+        return this._halfExtents;
+    }
+
+    /**
+     * @attribute
+     * @type {Color}
+     */
+    set colorX(value) {
+        this._colorX.copy(value);
+        this._set('color_x', value);
+    }
+
+    get colorX() {
+        return this._colorX;
+    }
+
+    /**
+     * @attribute
+     * @type {Color}
+     */
+    set colorZ(value) {
+        this._colorZ.copy(value);
+        this._set('color_x', value);
+    }
+
+    get colorZ() {
+        return this._colorZ;
+    }
+
+    /**
+     * @param {CameraComponent} camera - The camera component.
+     */
+    attach(camera) {
+        this._camera = camera;
+
+        this._prerender = () => {
```

Comment:
any reason this prerender call does not execute inside onPreRenderLayer? It belongs there just before the drawing itself, to guarantee no other code can change it inbetween.